packages/core/src/utils/storage/google-storage.ts

```diff
@@ -0,0 +1,33 @@
+import { Storage } from '@google-cloud/storage';
+
+import type { UploadFile } from './types.js';
+
+const defaultPublicDomain = 'storage.googleapis.com';
+
+export const buildGoogleStorage = (projectId: string, keyFilename: string, bucketName: string) => {
+  const storage = new Storage({ projectId, keyFilename });
+  const bucket = storage.bucket(bucketName);
+
+  const uploadFile: UploadFile = async (
+    data: Buffer,
+    objectKey: string,
+    { contentType, publicUrl } = {}
+  ) => {
+    const file = bucket.file(objectKey);
+
+    await file.save(data, {
+      contentType,
+      public: Boolean(publicUrl),
+    });
+
+    if (publicUrl) {
+      return { url: `${publicUrl}/${objectKey}` };
```

Comment:
We have a `appendPath` util method in our `@logto/essentials` package. You can just use it for concatenating URLs safely. E.g.
```ts
return { url: appendPath(publicUrl, objectKey) };
```

Reply:
To update this i would have to update the types 
`type UploadFileOptions = {
  contentType?: string;
  **publicUrl?: string;**
};

export type UploadFile = (
  data: Buffer,
  objectKey: string,
  options?: UploadFileOptions
) => **Promise<{ url: string }>;**
`
Then i would also have to make similar changes in the azure and s3 files. So should i do all of these changes ?

Reply:
Oh, I forgot about the type. But you don't have to update the types, you can `appendPath` first and `toString`.

E.g.
```ts
return { url: appendPath(publicUrl, objectKey).toString() }; // Or `.href`
```